### PR TITLE
This pull request implements the game board both in Go (for simulation and testing purposes) and in Solidity

### PR DESCRIPTION
**Pull request by zomglings**
Game board implementation in Solidity and Go.

### Solidity

- Implements `DegenTrail` contract which has several methods concerning the game board.
- The `DegenTrail` constructor configures the state of the game board on the 0th column.
- `DegenTrail.board` is a view method that returns the state of the game board at a given list of hexes.

### Go

- Introduces the `trailmix` CLI to interact with and analyze the game.
- `trailmix board` command which simulates board exploration and renders as SVG.

### CODE REVIEW

File: src/Game.sol [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    /// @param rerollFee Fee for re-roll on any action, assuming player doesn't want to accept their fate.
+    constructor(uint256 blocksToAct, uint256 rollFee, uint256 rerollFee)
+        Bandit(blocksToAct, address(this), rollFee, rerollFee)
+        ERC20("Supply", "SUPPLY")
+    {
+        uint256 prevBlockNumber = 0;
+        if (block.number > 0) {
+            prevBlockNumber = block.number - 1;
+        }
+        // Zero out the leading 12 bits of the block hash, to prevent overflows when adding 31 * j.
```
Comment (kellan-simiotics) [inline src/Game.sol]: Wouldn't zeroing 1 bit accomplish this?

Comment (kellan-simiotics) [reply src/Game.sol]: Oh, I think just the comment is wrong ... right shift then left shift zeroes the trailing bits, correct?

File: test/Game.t.sol [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+            uint256 terrainType = states[k][2] >> 1;
+
+            uint8 hexEntropy = uint8((startingEntropy + (31 * k)) & 0x7F);
+            assertLt(hexEntropy, game.EnvironmentDistributions(0, terrainType));
+            if (terrainType < 6) {
+                assertGe(hexEntropy, game.EnvironmentDistributions(0, terrainType + 1));
+            }
+        }
+    }
```
Comment (kellan-simiotics) [inline test/Game.t.sol:96]: Should we add tests that _explore correctly updates the board?

Comment (kellan-simiotics) [reply test/Game.t.sol:96]: Is it difficult to test internal method?

File: test/Game.t.sol [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        uint256 startingEntropy = uint256(blockhash(priorBlockNumber)) >> 12 << 12;
+
+        for (uint256 k = 0; k < 100; k++) {
+            assertEq(states[k][0], 0);
+            assertEq(states[k][1], 2 * k);
+            assertEq(states[k][2] % 2, 1);
+
+            uint256 terrainType = states[k][2] >> 1;
+
+            uint8 hexEntropy = uint8((startingEntropy + (31 * k)) & 0x7F);
```
Comment (kellan-simiotics) [inline test/Game.t.sol:90]: I think maybe we could use a comment here to explain that we are verifying the environment assignment for i=0 (forest) in a specific deployment.

**Review by kellan-simiotics — APPROVED** [review]
lg

File: test/Game.t.sol (hunk 2 above)
Comment (zomglings) [reply test/Game.t.sol:96]: Yeah, if we want to test the internal method, we should subclass in the tests to expose it publicly and test.

I'm planning to add those tests once we have wagons moving across the game board.

File: src/Game.sol (hunk 1 above)
Comment (zomglings) [reply src/Game.sol]: You are right, I just need to zero out the leading bit. I did `<< 1 >> 1`

File: src/Game.sol [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+            // 0011
+            Hex[i][j] = 3;
+        } else {
+            // 0001
+            Hex[i][j] = 1;
+        }
+    }
+
+    /// @notice Describes the environment of a hex with the given j-coordinate.
+    function environment(uint256 i) public pure returns (uint256) {
```
Comment (kellan-simiotics) [inline src/Game.sol:83]: I don't see any tests for the environment assignments. I don't know that we need them, but maybe a few comments about the flow of the environment and what this calculation is doing? Comment that the environment is constant for 32 column blocks?

**Review by kellan-simiotics — APPROVED** [review]

File: test/Game.t.sol (hunk 3 above)
Comment (zomglings) [reply test/Game.t.sol:90]: Done.

File: src/Game.sol (hunk 4 above)
Comment (zomglings) [reply src/Game.sol:83]: In Game.t.sol:

image

**Pull request merged**